### ISSUE TITLE
Jspm 0.16.25 module {x} not present when integrating a jspm bundle in another

### ISSUE THREAD

**Issue by hnb2** [jspm/jspm-cli#1555]
Hi,

I am currently integrating one of my jspm bundle A (under the form of an AMD SFX bundle) into another jspm project B.

But, I am running into the issue that the ids (e.g: define("**2**", ["require", "exports", "**1**"])... ) are colliding when making an AMD SFX bundle out of project B (which therefore include A). There could be twice the id '2' for instance.

Is it possible to keep the original module names when bundling (e.g: define("**MyModule**", ["require", "exports", "**MyOtherModule**"])... ) ? That way I could avoid those collisions. Unless there is another way to do this, please let me know, thanks.

**Comment by hnb2** [comment]
My issue is actually related to those tickets:

 * https://github.com/jspm/jspm-cli/issues/738
 * https://github.com/systemjs/builder/issues/168

Because the name of the AMD module is not set:

```js
.....
// End of projectA.min.js
(function(factory) {
define([], factory);
});
```
instead of:

```js
// How the end of projectA.min.js should be:
(function(factory) {
define("26", [], factory); // "26" here being the generated ID
});
```

I get the following error in the console : Module "26" not present.

I still haven't found how to set said name, and when I look at the [template](https://github.com/systemjs/builder/blob/master/templates/sfx-amd.js) I don't see how. Can I have some information about this ?

**Comment by guybedford** [comment]
Can you clarify what exactly you mean that the names are colliding? Does your build of projectA.min.js contain multiple `define("2"` statements? That would be a bug.

Or is the collission only happening when running a build in project B? If so, how is that collission happening? If you could share a build output or error log that would help to understand.

It should be impossible for both of the above cases to happen. As for the proposed issue in your second comment, that is the output AMD define, which is unrelated to the internal module rewrites entirely.

You can run a build without the name mapping by setting `encodeNames: false` as a build option or `--skip-encode-names` in jspm 0.17.

**Comment by hnb2** [comment]
Hi,

So I found the ‘encodeNames’ options in this [PR](https://github.com/systemjs/builder/pull/443/files) not so long before I received your comment.

So, l have the following situation:

 * In my project A’s package.json I define my main file as the AMD minified js file (done through jspm bundle-sfx)
 * I integrate my project A through jspm-git into projectB, dependencies are resolved, the name is inside map{} in jspm.conf.js. Everything is fine
 * I import components of my project A using ‘import’ statements in project B, my app is working fine in ‘development mode’ (calling the main file directly).
 * I bundle projectB using jspm-builder (gulp script)
 * When launching my app in ‘production mode’ (static bundle) I get: uncaught exception: Module 26 not present.

The final bundle looks like this:

```js 
// More self calling functions containing define(id, [a, b, c], fn)….
})(function(factory) {
  define([], factory); // This define should have the id “26”
);

_removeDefine();
   })();
   $__System.register(“e”, [“26”], function($__export) { //Note that “26” cannot be found anywhere
      return {
        setters: [function(m) {
          for (var p in m)
            $__export(p, m[p]);
        }],
        execute: function() {}
      };
   });

(function() {                                                                                                                                                                                                                                                                                                                                                          var _removeDefine = $__System.get("@@amd-helpers").createDefine();                                                                                                                                                                                                                                                                                                     define(“k”, ["require", "exports", “e”], function(require, exports, my_lib) { // Therefore “e” cannot be resolved
```

**Comment by guybedford** [comment]
Thanks that clarifies. So the problem is that the AMD build is not getting named correctly in subsequent builds. I will look into this.

**Comment by guybedford** [comment]
If you search the build source, are you sure you can't see another `define("26")` anywhere? We recently had another issue where another define is being named instead of the correct one.

**Comment by guybedford** [comment]
Also if it is possible to share your projectA.js SFX built output that would be a great help to test and resolve this.

**Comment by hnb2** [comment]
Well, this is the only occurent, which is kinda weird as it feels like it does not belong here, it can be found in the "middle" of a registerDynamic:

```js
   1592       if (typeof module === 'object' && typeof module.exports === 'object') {
   1593         module.exports = param;                                                                                                                                                 1594       } else if (typeof define === 'function' && define.amd) {                                                                                                                  1595         define(“26”, [], function() {                                                                                             1596           return param;                                                                                                                                                         1597         });                                                                                                                                                                     1598       } else {                                                                                                                                                                  1599         global.param = param;                                                                                                                                                   1600       }
```

And unfortunately I cannot share my project A as it is confidential material. But on my free time I could try to make a dummy project A.

**Comment by guybedford** [comment]
Perfect, then it is the same bug. I'm working on a fix now.

**Comment by hnb2** [comment]
Great, can't wait to apply your fix :)

**Comment by guybedford** [comment]
Patch in https://github.com/systemjs/builder/commit/e63c354feab6315ff36aa3e7be235cffc684c7c9. If you're able to try it out let me know!

**Comment by hnb2** [comment]
Silly question, but, to apply your patch, is it only by modifying the file: ```node_modules/systemjs-builder/compilers/amd.js``` in my project A and B ? Or can I force the version to master somewhere, in order to have all your other fixes ?

**Comment by guybedford** [comment]
In project B.

**Comment by hnb2** [comment]
Ok, so I modified node_modules/systemjs-builder/compilers/amd.js (commented the line in your patch) in project B and ran a bundle-sfx, but I still have the same output unfortunately.

**Comment by hnb2** [comment]
I modified as well node_modules/systemjs-builder/templates/amd-helper.js and ran minify.js from your [patch](https://github.com/systemjs/builder/commit/3e81fc2b4e63ba98f90de7bb872f635c55c16265). But I still have the same error.

**Comment by hnb2** [comment]
In the end I wrote a small piece of 'glue' in my gulp build task:

```js
        builder.loadConfig(path.join(config.projectDir, 'jspm.conf.js'))
            .then(() => {
                const moduleName = 'app';

                const buildConfig = {
                    format: 'amd',
                    minify: false, // Important for the issue #1555
                    mange: false, // Important for the issue #1555
                    sourceMaps: true
                };
                return builder.buildStatic(moduleName, outFile, buildConfig);
            })
            .then((output) => {
                // Walk around this systemjs-builder issue
                // https://github.com/jspm/jspm-cli/issues/1555
                output.source = output.source.replace(
                    'define([], factory)',
                    'define("registry:group/projectA@master/dist/myfile.min.js",[],factory)'
                );
                fs = require('fs');
                fs.writeFile('dist/' + distFileName, output.source, 'utf8');
         });
```

Then I changed the build configuration of Project B to never encode names (encodeNames: false), and everything is working fine in both development and production mode. It's extremely hack-ish and rely on names and version, but I could not figure out anything better than that.

**Comment by hnb2** [comment]
So, I took the time to create 2 dummy projects to reproduce the error (uncaught exception: Module 2 not present.), you can find them here:

 * https://github.com/hnb2/jspm-project-a
 * https://github.com/hnb2/jspm-project-b

Please let me know if you run into any installation related issues, thanks.

**Comment by og-neth** [comment]
@guybedford I left a comment about a similar issue I am getting on my static app.min.js bundle after running a systemjs build grunt task. The error is:

`Uncaught Module 10 not present.`

Here is my comment

[https://github.com/systemjs/builder/issues/167#issuecomment-188540098](https://github.com/systemjs/builder/issues/167#issuecomment-188540098)

I think this is inline with @hnb2 explanation that a previously defined AMD minified module (like Kendo or other kendo modules) is getting it's module name re-written. 

I checked my bundle and the module "10" is not defined anywhere, but it's getting re-written here:

`$__System.register("14",["2","4","3","d","f","10","13"],`

**Comment by Xample** [comment]
+1 same here i.e. importing a jspm bundle into another jspm project not working.

**Comment by guybedford** [comment]
This is now implemented in https://github.com/systemjs/builder/commit/69fe75ffd039307e5ad9a882ad1ae42323cd073f and https://github.com/systemjs/builder/commit/990eb8b6b490e7b058af0c576dfee6575fc5ba30 pending release.

**Comment by hnb2** [comment]
Hi @guybedford, so I did the following to test your patch, let me know if I did something wrong:

 * ```cd jspm-project-b```
 * ```cd node_modules/```
 * ```mv systemjs-builder systemjs-builder.old```
 * ```git clone https://github.com/systemjs/builder.git systemjs-builder```
 *  ```cd ..```
 * ```jspm bundle-sfx src/app.js dist/app.min.js --format global```

Then I opened dist/index.html and got the following error: ```uncaught exception: Module 3 not present.```

```
jspm --version
0.16.29
Running against local jspm install.
```

**Comment by guybedford** [comment]
@hnb2 you'd need to first recreate the bundle from project-a for this fix to work.

**Comment by hnb2** [comment]
Seems like it's working on my dummy projects !

I did the following:

 * cd jspm-project-a
 * //Same things to get your patch, as project B
 * // git add/commit/push
 * cd jspm-project-b
 * jspm install projecta=github:hnb2/jspm-project-a@master
 * jspm bundle-sfx src/app.js dist/app.min.js --format global

Now I will try on my real project and let you know !

By the way, when will you release this fix and for which release ? 16 or 17 ?

**Comment by guybedford** [comment]
Great to hear that and thanks for verifying. This will be on 0.16 as well as 0.17.

**Comment by hnb2** [comment]
That's fantastic, thanks a lot for fixing this !

**Comment by guybedford** [comment]
Released in 0.16.30.
